### PR TITLE
Typechecking

### PR DESCRIPTION
**Pull request by elyes298**
tackles issue #244 
I've completed the correction of the typing errors that come up when running ty typechecker. 
Also all the tests should work now

---
## Summary by cubic
Enable strict type checking with `ty` across the codebase and fail on warnings in pre-commit and CI. Tightens types, replaces broad ignores with safe casts/asserts, and adds small runtime checks without changing behavior.

- **Refactors**
  - Add `ty` pre-commit hook (`uvx ty check --error-on-warning`) and CI; configure `[tool.ty.src]` to exclude `tests`; add `chex` to `dev`.
  - Replace broad ignores with targeted casts/asserts across utils (`check_specs`, `check_shape_dtype`), pytrees, devices/logger, and plotting; annotate JSON export types; avoid private JAX sharding internals.
  - Accept `ArrayLike` for mode permittivity/permeability and coerce via `jnp.asarray`; also coerce arrays and cast extents in `plot_material`.
  - Unify placement constraints via `AnyConstraint`; switch APIs to `Sequence`; update call sites (e.g., `object_list=[volume]`).
  - Plotting: use `Color.to_mpl()`, assert `Figure`/`Axes`, and cast extents/limits; normalize 9‑tuple material properties.

- **Bug Fixes**
  - Stopping conditions: require `setup()`, validate `_spp`/`min_steps`/`max_steps`, clip dynamic slices, use locals for JIT safety, and raise clear errors if uninitialized.
  - Energy computation: detect anisotropy via shape checks to expand tensors correctly.
  - Waveguide/mode checks: pass `object_list=[volume]`; ensure array conversions.

<sup>Written for commit 6a37718876e0ffbea0d2057dc3ac1159e5ebe3d8. Summary will update on new commits.</sup>

### CODE REVIEW

**Comment by l-berg** [comment]
This is truly a far-reaching task. But it looks like there are still some type errors, right?

File: src/fdtdx/core/jax/sharding.py [hunk 1]
```diff
@@ -47,8 +47,9 @@ def get_dtype_bytes(dtype: jnp.dtype) -> int:
 def pretty_print_sharding(sharding: jax.sharding.Sharding) -> str:
     if isinstance(sharding, jax.sharding.NamedSharding):
         return f"NamedSharding({sharding.mesh.devices}, {sharding.spec})"
-    elif isinstance(sharding, jax.sharding.PositionalSharding):
-        return f"PositionalSharding({sharding._devices}, {sharding._ids.shape})"
+    #
+    # elif isinstance(sharding, jax.sharding.PositionalSharding):
+    #    return f"PositionalSharding({sharding._devices}, {sharding._ids.shape})"
```
Comment (ymahlau) [inline src/fdtdx/core/jax/sharding.py:53]: what is the reason for uncommenting these lines?

File: src/fdtdx/core/plotting/device_permittivity_index_utils.py [hunk 2]
```diff
@@ -55,8 +53,9 @@ def device_matrix_index_figure(
         - Proper axis labels and grid settings
     """
     assert device_matrix_indices.ndim == 3
-
-    fig, ax = cast(tuple[Figure, Axes], plt.subplots(figsize=(12, 12)))
+    fig: Figure
+    ax: Axes
+    fig, ax = plt.subplots(figsize=(12, 12))
```
Comment (ymahlau) [inline src/fdtdx/core/plotting/device_permittivity_index_utils.py:56]: I think a better way to do this would be an `assert isinstance(fig, Figure)` after the creation. Same for ax.

File: src/fdtdx/fdtd/initialization.py [hunk 3]
```diff
@@ -612,9 +612,14 @@ def _init_params(
     return params
 
 
+AnyConstraint = (
```
[the remark this answers is not in the document]
Comment (ymahlau) [reply src/fdtdx/fdtd/initialization.py]: @elyes298 ai review is correct here, please change

File: src/fdtdx/fdtd/stop_conditions.py [hunk 4]
```diff
@@ -133,6 +133,8 @@ def __call__(self, state: SimulationState, config: SimulationConfig, objects: Ob
         Returns:
             jax.Array: Boolean scalar - True if condition not met and curr_time_step < max_steps.
         """
+        if self.max_steps is None or self.min_steps is None:
+            raise RuntimeError("EnergyThresholdCondition.setup() must be calle before use. ")
```
Comment (ymahlau) [inline src/fdtdx/fdtd/stop_conditions.py]: typo

File: src/fdtdx/objects/device/device.py [hunk 5]
```diff
@@ -213,10 +213,11 @@ def __call__(
         # walk through modules
         for transform in self.param_transforms:
             check_specs(params, transform._input_shape)
-            params = transform(params, **transform_kwargs)
+            params_dict = cast(dict[str, jax.Array], params)
+            params = transform(params_dict, **transform_kwargs)
             check_specs(params, transform._output_shape)
         if len(params) == 1:
-            params = list(params.values())[0]
+            params = cast(jax.Array, list(params.values())[0])
```
Comment (ymahlau) [inline src/fdtdx/objects/device/device.py]: again, for single class checks (no composites like dict etc.) I would prefer to use `assert isinstance(...., jax.Array)`

File: src/fdtdx/utils/plot_setup.py [hunk 6]
```diff
@@ -135,7 +135,8 @@ def plot_setup_from_side(
                     if not o.name.startswith("Object"):
                         print_single = True
             label = o.__class__.__name__ if o.name.startswith("Object") else o.name
-            patch = Patch(color=o.color, label=label)
+            color_val = str(o.color)
```
[the remark this answers is not in the document]
Comment (ymahlau) [reply src/fdtdx/utils/plot_setup.py]: @elyes298 please take a look at this

File: src/fdtdx/core/jax/sharding.py (hunk 1 above)
Comment (elyes298) [reply src/fdtdx/core/jax/sharding.py:53]: @ymahlau I commented these because PositionalSharding was removed in recent versions and running it would cause an AttributeError. I can just delete these lines altogether

File: docs/source/advanced.rst [hunk 7]
```diff
@@ -7,6 +7,7 @@ Advanced
 
    notebooks/modes.ipynb
    notebooks/simulate_lossy_material.ipynb
+   notebooks/2D-Simulation.ipynb
```
[the remark this answers is not in the document]
Comment (elyes298) [reply docs/source/advanced.rst:10]: False positive: The notebooks are downloaded from the fdtdx-notebooks repository dynamically during the CI pipeline, so they aren't present in this repo for static analysis.
